src/residuals.cc

```diff
@@ -0,0 +1,155 @@
+    for (int i=0; i < nsamples; i++) {
+        auto sample_name = string(bcf_hdr_int2id(hdr, BCF_DT_SAMPLE, i));
+        samples.push_back(sample_name);
+    }
+
+    return std::accumulate(samples.begin(), samples.end(), std::string(),
+                    [](const std::string& a, const std::string& b) -> std::string {
+                        return a + (a.length() > 0 ? "," : "") + b;
+                    } );
+}
+
+// destructor
+Residuals::~Residuals() {}
+
+Status Residuals::Open(const MetadataCache& cache, BCFData& data,
+                       const std::string& sampleset, const std::vector<std::string>& samples,
+                       unique_ptr<Residuals> &ans) {
+    ans = make_unique<Residuals>(cache, data, sampleset, samples);
+    return Status::OK();
+}
+
+
+Status Residuals::gen_record(const unified_site& site,
+                             const bcf_hdr_t *gl_hdr,
+                             const bcf1_t *gl_call,
+                             std::string &ynode) {
+    Status s;
+    YAML::Emitter out;
+
+    out << YAML::BeginMap;
+
+    // write the original records
+    out << YAML::Key << "input";
+    out << YAML::BeginMap;
+    out << YAML::Key << "body";
+    out << YAML::BeginSeq;
+    shared_ptr<const set<string>> samples2, datasets;
+    vector<unique_ptr<RangeBCFIterator>> iterators;
+    S(data_.sampleset_range(cache_, sampleset_, site.pos,
+                           samples2, datasets, iterators));
+    assert(samples_.size() == samples2->size());
+
+    // for each pertinent dataset
+    for (const auto& dataset : *datasets) {
+        // load BCF records overlapping the site by "merging" the iterators
+        shared_ptr<const bcf_hdr_t> dataset_header;
+        vector<vector<shared_ptr<bcf1_t>>> recordss(iterators.size());
+        vector<shared_ptr<bcf1_t>> records;
+
+        for (const auto& iter : iterators) {
+            string this_dataset;
+            vector<shared_ptr<bcf1_t>> these_records;
+            S(iter->next(this_dataset, dataset_header, these_records));
+            if (dataset != this_dataset) {
+                return Status::Failure("genotype_site: iterator returned unexpected dataset",
+                                       this_dataset + " instead of " + dataset);
+            }
+            records.insert(records.end(), these_records.begin(), these_records.end());
+        }
+
+        for (auto rec : records) {
+            out << YAML::BeginMap;
+            out << YAML::Key << dataset;
+            string samples = samples_from_dataset(dataset_header.get());
+            out << YAML::Literal << samples + "\n" + *(bcf1_to_string(dataset_header.get(), rec.get()));
+            out << YAML::EndMap;
+        }
+    }
+    out << YAML::EndSeq;
+    out << YAML::EndMap;
+
+    // write the unified site
+    out << YAML::Key << "unified_site";
+    const auto& contigs = cache_.contigs();
+    S(site.yaml(contigs, out));
+
+
+    // write the output bcf, the calls that GLnexus made.
```

Comment:
The residuals file needs to specify the order of the samples in the output VCF. That's crucial because it's not in general the same as the order in the inputs.